### DISCUSSION TITLE
Add selectwidget attribute for option via typing?

### DISCUSSION THREAD

**Discussion by twMat** [TiddlyWiki/TiddlyWiki5#5508] [Ideas]
Is this ready as Issue proposal?

I occasionally need a selectwidget... but that behaves like the tag picker or the search field, where you can select an item OR optionally manually type a selection.

I envision an enhanced SelectWidget with a new attribute, e.g `option=type` or `option=text` or some such:

```
<$select tiddler='$:/SiteTitle' option='type'>
<option value='cities'>A Tale of Two Cities</option>
<option value='science'>A New Kind of Science</option>
<option value='dice'>The Dice Man</option>
</$select>
```

Makes sense?

**Comment by Jermolene** [comment]
Hi @twMat sadly HTML doesn't (yet) support what describe (Windows calls it a combobox). That's why the core constructs it from primitive wikitext components. We could do a better job of making them reusable, but that's probably a matter of macros/templates rather than changing `<$select>`.

**Reply by twMat** [reply]
Ah, combobox, right.

How about introducing the html **datalist** element! I would guess that this had too poor CanIUse perviously but now it's [95%](https://caniuse.com/?search=datalist). Here's an [example](https://stackoverflow.com/questions/49186286/plain-html-combo-box-code).

**Reply by Jermolene** [reply]
Hi @twMat doh! I'd forgotten about that, it's a relatively recent addition. However, frustratingly, the way that the `<datalist>` element is attached to the `<input>` element is via an ID. From the example you quoted:

```
<label for="countries">Choose a Country:</label>
<input list="countryList" id="countries" name="countries" />
<datalist id="countryList">
  <option value="India">
  <option value="France">
  <option value="Isreal">
  <option value="USA">
  </datalist>
```

That's a problem for TW5 because of the requirement that IDs be unique across the document. Without using `<<qualify>>` or some other mechanism it wouldn't be possible to have multiple comboboxes displayed at once.

**Reply by twMat** [reply]
>That's a problem for TW5 because of the requirement that IDs be unique across the document. Without using <<qualify>> or some other mechanism it wouldn't be possible to have multiple comboboxes displayed at once.

I assumed that in the transition from the html element into a TW widget, this was the kind of things that are added? Or are the html based widgets just a kind of empty wrap without modifications to the html tag functionality?

**Reply by Jermolene** [reply]
> I assumed that in the transition from the html element into a TW widget, this was the kind of things that are added? Or are the html based widgets just a kind of empty wrap without modifications to the html tag functionality?

No, it's not a problem we have to address at the moment because the core doesn't depend on any functionality that requires IDs.

But in any event, the HTML 5 control is still handled somewhat inconsistently between browsers, and has limited styling options, so I don't think it's a great loss. We couldn't use it for our tags dropdowns, for example.

I still think the best answer to the OP is to make our existing combobox be easier to reuse.

**Reply by twMat** [reply]
>I still think the best answer to the OP is to make our existing combobox be easier to reuse.

Is this "packaged" as a macro anywhere in the code? And is it the same code (macro) used for the various things that do use something like this? (tag dropdown, search field, (more?)). If "yes", I'll search for it myself.

But, yes, it would be neat if this was made as an official macro or something.

Thanks for the replies!
